File: go/events.go
package nitroping

import (
	"context"
	"errors"
)

// EventsService wraps the /api/v1/events endpoint — the public,
// unauthenticated engagement endpoint. The (NotificationID, DeviceID)
// pair is the bearer secret, so no Authorization header is strictly
// required (a pk_ public key is also fine). This is what a client app
// calls when a notification is opened or an action is clicked.
//
// Construct one via NewClient.
type EventsService struct {
	transport *transport
}

// Report reports an engagement event (EventOpened or EventClicked).
// NotificationID, DeviceID and Type are required; ActionID and
// HappenedAt are optional. On success the result is {Accepted: true}.
//
// Returns *APIError with Code="not_found" (404) if the notification /
// device pair is unknown.
//
//	res, err := client.Events.Report(ctx, nitroping.ReportEventRequest{
//	    NotificationID: "n1",
//	    DeviceID:       "d1",
//	    Type:           nitroping.EventOpened,
//	})
func (s *EventsService) Report(
	ctx context.Context,
	req ReportEventRequest,
	opts ...RequestOption,
) (*EventResult, error) {
	if req.NotificationID == "" {
		return nil, errors.New("nitroping: notification id is required")
	}
	if req.DeviceID == "" {
		return nil, errors.New("nitroping: device id is required")
	}
	cfg := applyOptions(opts)
	var out EventResult
	if err := s.transport.do(ctx, "POST", "/api/v1/events", req, cfg, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: js/src/events.ts
/**
 * `events` resource client.
 *
 * Mounted on `Nitroping` as `np.events`. Wraps `POST /api/v1/events` —
 * the public, unauthenticated engagement endpoint. The
 * `(notificationId, deviceId)` pair is the bearer secret, so no
 * `Authorization` header is required (and a `pk_` public key is fine).
 *
 * This is the endpoint a client app calls when a notification is opened
 * or a notification action is clicked.
 */

import type { HttpClient } from "./http"
import type { ReportEventRequest } from "./types"

export class EventsClient {
  constructor(private readonly http: HttpClient) {}

  /**
   * Report an engagement event (`opened` or `clicked`).
   *
   * Resolves to `{ accepted: true }` on 202. Throws a `NitropingError`
   * with `code: "not_found"` (404) if the notification/device pair is
   * unknown.
   */
  async report(input: ReportEventRequest): Promise<{ accepted: boolean }> {
    const body: Record<string, unknown> = {
      notification_id: input.notificationId,
      device_id: input.deviceId,
      type: input.type,
    }
    if (input.actionId !== undefined) body["action_id"] = input.actionId
    if (input.happenedAt !== undefined) body["happened_at"] = input.happenedAt

    return await this.http.request<{ accepted: boolean }>("POST", "/api/v1/events", { body })
  }
}
